File: aac/api/python/wia_aac/output/braille_adapter.py
"""
WIA AAC Braille Adapter
Phase 4: WIA Ecosystem Integration

WIA Braille integration for braille output
"""

import asyncio
from abc import abstractmethod
from typing import Optional, List, Dict

from .base_output import BaseOutputAdapter
from .types import (
    OutputType,
    OutputState,
    OutputOptions,
    OutputEventType,
    BrailleOutput,
    BrailleDisplay,
    OutputError,
    OutputErrorCode
)


class IBrailleAdapter(BaseOutputAdapter):
    """Braille adapter interface"""

    @abstractmethod
    async def text_to_ipa(self, text: str) -> str:
        """Convert text to IPA"""
        pass

    @abstractmethod
    async def text_to_braille(self, text: str) -> BrailleOutput:
        """Convert text to braille"""
        pass

    @abstractmethod
    async def send_to_display(self, braille: BrailleOutput) -> None:
        """Send braille to display"""
        pass

    @abstractmethod
    async def get_connected_displays(self) -> List[BrailleDisplay]:
        """Get connected displays"""
        pass

    @abstractmethod
    def set_display(self, display_id: str) -> None:
        """Set active display"""
        pass


class MockBrailleAdapter(IBrailleAdapter):
    """Mock Braille Adapter using WIA Braille system (IPA-based)"""

    def __init__(self):
        super().__init__()
        self._ipa_map: Dict[str, str] = {}
        self._braille_map: Dict[str, str] = {}
        self._selected_display_id: Optional[str] = None
        self._mock_displays: List[BrailleDisplay] = []

    @property
    def type(self) -> OutputType:
        return OutputType.BRAILLE

    @property
    def name(self) -> str:
        return "MockBraille"

    async def initialize(self, options: Optional[OutputOptions] = None) -> None:
        """Initialize the adapter"""
        self._load_ipa_map()
        self._load_braille_map()
        self._init_mock_displays()
        self._state = OutputState.IDLE
        self._emit(OutputEventType.START, {"message": "Braille adapter initialized"})

    def _init_mock_displays(self) -> None:
        """Initialize mock braille displays"""
        self._mock_displays = [
            BrailleDisplay(id="mock-40", name="Mock Braille Display 40", cells=40, connected=True),
            BrailleDisplay(id="mock-80", name="Mock Braille Display 80", cells=80, connected=False)
        ]
        self._selected_display_id = "mock-40"

    def _load_ipa_map(self) -> None:
        """Load IPA mapping for common words"""
        # Korean words
        self._ipa_map["안녕"] = "/annjʌŋ/"
        self._ipa_map["안녕하세요"] = "/annjʌŋhasejo/"
        self._ipa_map["감사"] = "/kamsa/"
        self._ipa_map["감사합니다"] = "/kamsahamnida/"
        self._ipa_map["사랑"] = "/saraŋ/"
        self._ipa_map["사랑해"] = "/saraŋhɛ/"
        self._ipa_map["네"] = "/ne/"
        self._ipa_map["아니오"] = "/anio/"
        self._ipa_map["도움"] = "/toum/"
        self._ipa_map["미안"] = "/mian/"
        self._ipa_map["좋아"] = "/tʃoa/"

        # English words
        self._ipa_map["hello"] = "/həˈloʊ/"
        self._ipa_map["thanks"] = "/θæŋks/"
        self._ipa_map["love"] = "/lʌv/"
        self._ipa_map["help"] = "/hɛlp/"
        self._ipa_map["yes"] = "/jɛs/"
        self._ipa_map["no"] = "/noʊ/"

    def _load_braille_map(self) -> None:
        """Load IPA to Braille mapping (WIA Braille system - 8-dot)"""
        # Vowels
        self._braille_map["a"] = "⠁"  # U+2801
        self._braille_map["e"] = "⠑"  # U+2811
        self._braille_map["i"] = "⠊"  # U+280A
        self._braille_map["o"] = "⠕"  # U+2815
        self._braille_map["u"] = "⠥"  # U+2825
        self._braille_map["ə"] = "⠢"  # U+2822 (schwa)
        self._braille_map["ʌ"] = "⠪"  # U+282A
        self._braille_map["ɛ"] = "⠫"  # U+282B
        self._braille_map["ɔ"] = "⠬"  # U+282C
        self._braille_map["æ"] = "⠭"  # U+282D
        self._braille_map["ʊ"] = "⠮"  # U+282E

        # Consonants
        self._braille_map["p"] = "⠏"  # U+280F
        self._braille_map["b"] = "⠃"  # U+2803
        self._braille_map["t"] = "⠞"  # U+281E
        self._braille_map["d"] = "⠙"  # U+2819
        self._braille_map["k"] = "⠅"  # U+2805
        self._braille_map["g"] = "⠛"  # U+281B
        self._braille_map["m"] = "⠍"  # U+280D
        self._braille_map["n"] = "⠝"  # U+281D
        self._braille_map["ŋ"] = "⠻"  # U+283B (eng)
        self._braille_map["f"] = "⠋"  # U+280B
        self._braille_map["v"] = "⠧"  # U+2827
        self._braille_map["s"] = "⠎"  # U+280E
        self._braille_map["z"] = "⠵"  # U+2835
        self._braille_map["h"] = "⠓"  # U+2813
        self._braille_map["l"] = "⠇"  # U+2807
        self._braille_map["r"] = "⠗"  # U+2817
        self._braille_map["w"] = "⠺"  # U+283A
        self._braille_map["j"] = "⠚"  # U+281A
        self._braille_map["ʃ"] = "⠱"  # U+2831 (sh)
        self._braille_map["ʒ"] = "⠴"  # U+2834 (zh)
        self._braille_map["tʃ"] = "⠹"  # U+2839 (ch)
        self._braille_map["θ"] = "⠹"  # U+2839 (th voiceless)
        self._braille_map["ð"] = "⠮"  # U+282E (th voiced)

        # Diacritics
        self._braille_map["ˈ"] = "⠄"  # U+2804 (primary stress)
        self._braille_map["ˌ"] = "⠠"  # U+2820 (secondary stress)
        self._braille_map["ː"] = "⠒"  # U+2812 (long)

        # Space
        self._braille_map[" "] = "⠀"  # U+2800 (blank)

    def _get_braille_char(self, ipa_char: str) -> str:
        """Get braille character for an IPA character"""
        return self._braille_map.get(ipa_char, "⠿")  # U+283F for unknown

    def _tokenize_ipa(self, ipa: str) -> List[str]:
        """Tokenize IPA string into phonemes"""
        tokens = []
        i = 0

        while i < len(ipa):
            # Check for two-character combinations
            if i < len(ipa) - 1:
                two_char = ipa[i:i+2]
                if two_char in self._braille_map:
                    tokens.append(two_char)
                    i += 2
                    continue

            # Single character
            tokens.append(ipa[i])
            i += 1

        return tokens

    async def text_to_ipa(self, text: str) -> str:
        """Convert text to IPA"""
        words = text.lower().split()
        ipa_words = []

        for word in words:
            ipa = self._ipa_map.get(word)
            if ipa:
                ipa_words.append(ipa)
            else:
                # Simple fallback
                ipa_words.append(f"/{word}/")

        return " ".join(ipa_words)

    async def text_to_braille(self, text: str) -> BrailleOutput:
        """Convert text to braille"""
        ipa = await self.text_to_ipa(text)

        # Remove IPA delimiters
        clean_ipa = ipa.replace("/", "").replace("[", "").replace("]", "")
        chars = self._tokenize_ipa(clean_ipa)

        braille = ""
        unicode_list = []
        dots = []

        for char in chars:
            braille_char = self._get_braille_char(char)
            braille += braille_char

            code_point = ord(braille_char)
            unicode_list.append(f"U+{code_point:04X}")
            dots.append(code_point - 0x2800)

        return BrailleOutput(
            text=text,
            ipa=ipa,
            braille=braille,
            unicode=unicode_list,
            dots=dots
        )

    async def send_to_display(self, braille: BrailleOutput) -> None:
        """Send braille to display"""
        if not self._selected_display_id:
            raise OutputError(
                OutputErrorCode.DEVICE_NOT_CONNECTED,
                "No braille display selected",
                True
            )

        display = next(
            (d for d in self._mock_displays if d.id == self._selected_display_id),
            None
        )

        if not display or not display.connected:
            raise OutputError(
                OutputErrorCode.DEVICE_NOT_CONNECTED,
                "Braille display not connected",
                True
            )

        self._state = OutputState.OUTPUTTING
        self._emit(OutputEventType.START, {
            "displayId": self._selected_display_id,
            "braille": braille.braille
        })

        # Simulate display reading time
        reading_time = max(len(braille.braille) * 0.1, 0.5)
        await asyncio.sleep(reading_time)

        self._state = OutputState.IDLE
        self._emit(OutputEventType.END, {
            "displayId": self._selected_display_id,
            "braille": braille.braille
        })

    async def output(self, text: str, options: Optional[OutputOptions] = None) -> None:
        """Output text as braille"""
        braille_output = await self.text_to_braille(text)
        await self.send_to_display(braille_output)

    async def get_connected_displays(self) -> List[BrailleDisplay]:
        """Get connected displays"""
        return [d for d in self._mock_displays if d.connected]

    def set_display(self, display_id: str) -> None:
        """Set active display"""
        display = next(
            (d for d in self._mock_displays if d.id == display_id),
            None
        )
        if not display:
            raise OutputError(
                OutputErrorCode.DEVICE_NOT_CONNECTED,
                f"Display not found: {display_id}",
                True
            )
        self._selected_display_id = display_id

    def stop(self) -> None:
        """Stop output"""
        self._state = OutputState.IDLE

    def is_available(self) -> bool:
        """Check if adapter is available"""
        return any(d.connected for d in self._mock_displays)

    async def dispose(self) -> None:
        """Dispose resources"""
        self._ipa_map.clear()
        self._braille_map.clear()
        self._handlers.clear()

    def add_ipa_mapping(self, word: str, ipa: str) -> None:
        """Add custom IPA mapping"""
        self._ipa_map[word.lower()] = ipa

    def add_braille_mapping(self, ipa: str, braille: str) -> None:
        """Add custom braille mapping"""
        self._braille_map[ipa] = braille

    def connect_display(self, display_id: str) -> None:
        """Simulate connecting a display"""
        for d in self._mock_displays:
            if d.id == display_id:
                d.connected = True
                break

    def disconnect_display(self, display_id: str) -> None:
        """Simulate disconnecting a display"""
        for d in self._mock_displays:
            if d.id == display_id:
                d.connected = False
                break

File: aac/api/typescript/src/output/BrailleAdapter.ts
/**
 * WIA AAC Braille Adapter
 * Phase 4: WIA Ecosystem Integration
 *
 * WIA Braille integration for braille output
 */

import { BaseOutputAdapter } from './IOutputAdapter';
import {
  OutputOptions,
  BrailleOutput,
  BrailleDisplay,
  OutputError,
  OutputErrorCode
} from './types';

/**
 * Braille adapter interface
 */
export interface IBrailleAdapter {
  /**
   * Convert text to IPA
   * @param text - Text to convert
   */
  textToIPA(text: string): Promise<string>;

  /**
   * Convert text to braille
   * @param text - Text to convert
   */
  textToBraille(text: string): Promise<BrailleOutput>;

  /**
   * Send braille to display
   * @param braille - Braille output to send
   */
  sendToDisplay(braille: BrailleOutput): Promise<void>;

  /**
   * Get connected displays
   */
  getConnectedDisplays(): Promise<BrailleDisplay[]>;

  /**
   * Set active display
   * @param displayId - Display ID
   */
  setDisplay(displayId: string): void;
}

/**
 * Mock Braille Adapter
 * Uses WIA Braille system (IPA-based)
 */
export class MockBrailleAdapter extends BaseOutputAdapter implements IBrailleAdapter {
  readonly type = 'braille' as const;
  readonly name = 'MockBraille';

  private ipaMap: Map<string, string> = new Map();
  private brailleMap: Map<string, string> = new Map();
  private selectedDisplayId: string | null = null;
  private mockDisplays: BrailleDisplay[] = [];

  async initialize(): Promise<void> {
    this.loadIPAMap();
    this.loadBrailleMap();
    this.initMockDisplays();
    this._state = 'idle';
    this.emit('start', { message: 'Braille adapter initialized' });
  }

  /**
   * Initialize mock braille displays
   */
  private initMockDisplays(): void {
    this.mockDisplays = [
      { id: 'mock-40', name: 'Mock Braille Display 40', cells: 40, connected: true },
      { id: 'mock-80', name: 'Mock Braille Display 80', cells: 80, connected: false }
    ];
    this.selectedDisplayId = 'mock-40';
  }

  /**
   * Load simple IPA mapping for common words
   * In production, this would use a proper G2P (Grapheme-to-Phoneme) library
   */
  private loadIPAMap(): void {
    // Korean words
    this.ipaMap.set('안녕', '/annjʌŋ/');
    this.ipaMap.set('안녕하세요', '/annjʌŋhasejo/');
    this.ipaMap.set('감사', '/kamsa/');
    this.ipaMap.set('감사합니다', '/kamsahamnida/');
    this.ipaMap.set('사랑', '/saraŋ/');
    this.ipaMap.set('사랑해', '/saraŋhɛ/');
    this.ipaMap.set('네', '/ne/');
    this.ipaMap.set('아니오', '/anio/');
    this.ipaMap.set('도움', '/toum/');
    this.ipaMap.set('미안', '/mian/');
    this.ipaMap.set('좋아', '/tʃoa/');

    // English words
    this.ipaMap.set('hello', '/həˈloʊ/');
    this.ipaMap.set('thanks', '/θæŋks/');
    this.ipaMap.set('love', '/lʌv/');
    this.ipaMap.set('help', '/hɛlp/');
    this.ipaMap.set('yes', '/jɛs/');
    this.ipaMap.set('no', '/noʊ/');
  }

  /**
   * Load IPA to Braille mapping (WIA Braille system - 8-dot)
   * Based on IPA phonemes to braille cells
   */
  private loadBrailleMap(): void {
    // Vowels
    this.brailleMap.set('a', '⠁');   // U+2801
    this.brailleMap.set('e', '⠑');   // U+2811
    this.brailleMap.set('i', '⠊');   // U+280A
    this.brailleMap.set('o', '⠕');   // U+2815
    this.brailleMap.set('u', '⠥');   // U+2825
    this.brailleMap.set('ə', '⠢');   // U+2822 (schwa)
    this.brailleMap.set('ʌ', '⠪');   // U+282A
    this.brailleMap.set('ɛ', '⠫');   // U+282B
    this.brailleMap.set('ɔ', '⠬');   // U+282C
    this.brailleMap.set('æ', '⠭');   // U+282D
    this.brailleMap.set('ʊ', '⠮');   // U+282E

    // Consonants
    this.brailleMap.set('p', '⠏');   // U+280F
    this.brailleMap.set('b', '⠃');   // U+2803
    this.brailleMap.set('t', '⠞');   // U+281E
    this.brailleMap.set('d', '⠙');   // U+2819
    this.brailleMap.set('k', '⠅');   // U+2805
    this.brailleMap.set('g', '⠛');   // U+281B
    this.brailleMap.set('m', '⠍');   // U+280D
    this.brailleMap.set('n', '⠝');   // U+281D
    this.brailleMap.set('ŋ', '⠻');   // U+283B (eng)
    this.brailleMap.set('f', '⠋');   // U+280B
    this.brailleMap.set('v', '⠧');   // U+2827
    this.brailleMap.set('s', '⠎');   // U+280E
    this.brailleMap.set('z', '⠵');   // U+2835
    this.brailleMap.set('h', '⠓');   // U+2813
    this.brailleMap.set('l', '⠇');   // U+2807
    this.brailleMap.set('r', '⠗');   // U+2817
    this.brailleMap.set('w', '⠺');   // U+283A
    this.brailleMap.set('j', '⠚');   // U+281A
    this.brailleMap.set('ʃ', '⠱');   // U+2831 (sh)
    this.brailleMap.set('ʒ', '⠴');   // U+2834 (zh)
    this.brailleMap.set('tʃ', '⠹');  // U+2839 (ch) - combined
    this.brailleMap.set('dʒ', '⠚');  // U+281A (j as in "judge")
    this.brailleMap.set('θ', '⠹');   // U+2839 (th voiceless)
    this.brailleMap.set('ð', '⠮');   // U+282E (th voiced)

    // Diacritics/Markers
    this.brailleMap.set('ˈ', '⠄');   // U+2804 (primary stress)
    this.brailleMap.set('ˌ', '⠠');   // U+2820 (secondary stress)
    this.brailleMap.set('ː', '⠒');   // U+2812 (long)

    // Space
    this.brailleMap.set(' ', '⠀');   // U+2800 (blank)
  }

  /**
   * Get braille character for an IPA character
   */
  private getBrailleChar(ipaChar: string): string {
    return this.brailleMap.get(ipaChar) || '⠿'; // U+283F for unknown
  }

  async textToIPA(text: string): Promise<string> {
    const words = text.toLowerCase().split(/\s+/);
    const ipaWords: string[] = [];

    for (const word of words) {
      const ipa = this.ipaMap.get(word);
      if (ipa) {
        ipaWords.push(ipa);
      } else {
        // Simple fallback: just wrap in slashes
        ipaWords.push(`/${word}/`);
      }
    }

    return ipaWords.join(' ');
  }

  async textToBraille(text: string): Promise<BrailleOutput> {
    const ipa = await this.textToIPA(text);

    // Remove IPA delimiters and process
    const cleanIPA = ipa.replace(/[\/\[\]]/g, '');
    const chars = this.tokenizeIPA(cleanIPA);

    let braille = '';
    const unicode: string[] = [];
    const dots: number[] = [];

    for (const char of chars) {
      const brailleChar = this.getBrailleChar(char);
      braille += brailleChar;

      const codePoint = brailleChar.charCodeAt(0);
      unicode.push(`U+${codePoint.toString(16).toUpperCase().padStart(4, '0')}`);
      dots.push(codePoint - 0x2800);
    }

    return {
      text,
      ipa,
      braille,
      unicode,
      dots
    };
  }

  /**
   * Tokenize IPA string into individual phonemes
   */
  private tokenizeIPA(ipa: string): string[] {
    const tokens: string[] = [];
    let i = 0;

    while (i < ipa.length) {
      // Check for two-character combinations first
      if (i < ipa.length - 1) {
        const twoChar = ipa.substring(i, i + 2);
        if (this.brailleMap.has(twoChar)) {
          tokens.push(twoChar);
          i += 2;
          continue;
        }
      }

      // Single character
      const char = ipa[i];
      tokens.push(char);
      i++;
    }

    return tokens;
  }

  async sendToDisplay(braille: BrailleOutput): Promise<void> {
    if (!this.selectedDisplayId) {
      throw new OutputError(
        OutputErrorCode.DEVICE_NOT_CONNECTED,
        'No braille display selected',
        true
      );
    }

    const display = this.mockDisplays.find(d => d.id === this.selectedDisplayId);
    if (!display || !display.connected) {
      throw new OutputError(
        OutputErrorCode.DEVICE_NOT_CONNECTED,
        'Braille display not connected',
        true
      );
    }

    this._state = 'outputting';
    this.emit('start', {
      displayId: this.selectedDisplayId,
      braille: braille.braille
    });

    // Simulate sending to display
    // In production, this would use BrlAPI or similar

    // Display duration based on braille length
    const readingTime = Math.max(braille.braille.length * 100, 500);
    await new Promise(resolve => setTimeout(resolve, readingTime));

    this._state = 'idle';
    this.emit('end', {
      displayId: this.selectedDisplayId,
      braille: braille.braille
    });
  }

  async output(text: string, options?: OutputOptions): Promise<void> {
    const brailleOutput = await this.textToBraille(text);
    await this.sendToDisplay(brailleOutput);
  }

  async getConnectedDisplays(): Promise<BrailleDisplay[]> {
    return this.mockDisplays.filter(d => d.connected);
  }

  setDisplay(displayId: string): void {
    const display = this.mockDisplays.find(d => d.id === displayId);
    if (!display) {
      throw new OutputError(
        OutputErrorCode.DEVICE_NOT_CONNECTED,
        `Display not found: ${displayId}`,
        true
      );
    }
    this.selectedDisplayId = displayId;
  }

  stop(): void {
    this._state = 'idle';
  }

  isAvailable(): boolean {
    return this.mockDisplays.some(d => d.connected);
  }

  async dispose(): Promise<void> {
    this.ipaMap.clear();
    this.brailleMap.clear();
    this.handlers.clear();
  }

  /**
   * Add custom IPA mapping
   */
  addIPAMapping(word: string, ipa: string): void {
    this.ipaMap.set(word.toLowerCase(), ipa);
  }

  /**
   * Add custom braille mapping
   */
  addBrailleMapping(ipa: string, braille: string): void {
    this.brailleMap.set(ipa, braille);
  }

  /**
   * Simulate connecting a display
   */
  connectDisplay(displayId: string): void {
    const display = this.mockDisplays.find(d => d.id === displayId);
    if (display) {
      display.connected = true;
    }
  }

  /**
   * Simulate disconnecting a display
   */
  disconnectDisplay(displayId: string): void {
    const display = this.mockDisplays.find(d => d.id === displayId);
    if (display) {
      display.connected = false;
    }
  }
}
